### PR TITLE
Implement release workflow in GitHub actions

### LINKED ISSUE #322: PyPi releases do not always match repo code
While trying to work out why installing `0.0.17` still triggered the issue outlined in #302, I started looking at other versions and discovered that a few of them don't appear to match the code tagged for that release in this repo.

I've done a rudimentary check on the last 5 versions by doing a very lazy test, just checking if the version of `pymysql` matches between the git tag and the pypi release, here are the results:

[`0.0.15`](https://github.com/aio-libs/aiomysql/blob/v0.0.15/setup.py#L7): match
```sh
$ docker container run -it --rm python pip install aiomysql==0.0.15 | grep -F 'Collecting PyMySQL'                                                                                                                 
Collecting PyMySQL>=0.7.5 (from aiomysql==0.0.15)
```

[`0.0.16`](https://github.com/aio-libs/aiomysql/blob/v0.0.16/setup.py#L7): doesn't match
```sh
$ docker container run -it --rm python pip install aiomysql==0.0.16 | grep -F 'Collecting PyMySQL'
Collecting PyMySQL<0.9,>=0.7.5 (from aiomysql==0.0.16)
```
[`0.0.17`](https://github.com/aio-libs/aiomysql/blob/v0.0.17/setup.py#L7): doesn't match
```sh
$ docker container run -it --rm python pip install aiomysql==0.0.17 | grep -F 'Collecting PyMySQL'
Collecting PyMySQL<=0.9.2,>=0.9 (from aiomysql==0.0.17)
```

[`0.0.18`](https://github.com/aio-libs/aiomysql/blob/v0.0.18/setup.py#L7): no release
```sh
$ docker container run -it --rm python pip install aiomysql==0.0.18 | grep -F 'Collecting PyMySQL'
err: No matching distribution found for aiomysql==0.0.18
```

[`0.0.19`](https://github.com/aio-libs/aiomysql/blob/v0.0.19/setup.py#L7): match
```sh
$ docker container run -it --rm python pip install aiomysql==0.0.19 | grep -F 'Collecting PyMySQL'
Collecting PyMySQL<=0.9.2,>=0.9 (from aiomysql==0.0.19)
```

At best this is awfully confusing because according to the code `0.0.17` was the first release to have the conservative version pinning for `pymysql`, but in fact that's some how made it into the `0.0.16` release (which the code doesn't reflect) but _not_ into `0.0.17`.

It's a little unsettling to find that the versions released in PyPi don't always match the tagged code releases. Is this some sort of mistake or quirk in the shipping process? 😕 

Possibly related: https://github.com/aio-libs/aiomysql/issues/318

### PR DESCRIPTION
**Pull request by Nothing4You**
## What do these changes do?

- heavily based on @webknjaz's work on https://github.com/cherrypy/cheroot/blob/b2c619f3e60682b9405a87cea48e8d30801b6048/.github/workflows/ci-cd.yml
- create package before running tests, run tests against same package that will be published
- run nightly tests
- publish package to test pypi for commits to master
- publish package to pypi using workflow_dispatch
- create github tag and release after successful tests on workflow_dispatch

## Are there changes in behavior for the user?

no

## Related issue number

fixes #322

## Checklist

- [x] I think the code is well written
- [ ] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [x] Add a new news fragment to `CHANGES.txt`

### CODE REVIEW

File: CHANGES.txt [hunk 1]
```diff
@@ -23,6 +23,7 @@ To be included in 1.0.0 (unreleased)
 * Fix error packet handling for SSCursor #428
 * Required python version is now properly documented in python_requires instead of failing on setup.py execution #731
 * Add rsa extras_require depending on PyMySQL[rsa] #557
+* Self-reported __version__ now returns version from installed package metadata, otherwise unknown #734
```
Comment (webknjaz) [inline CHANGES.txt:26]: How about
```suggestion
* Self-reported `__version__` now returns version from installed package metadata, otherwise `'unknown'` #734
```

File: MANIFEST.in [hunk 2]
```diff
@@ -1,5 +1,9 @@
...
-global-exclude *.pyc *.swp
+exclude .coveragerc
+exclude .flake8
+exclude .gitignore
+exclude docker-compose.yml
+exclude requirements-dev.txt
+exclude .github/**
+exclude docs/**
+exclude examples/**
+exclude tests/**
```
Comment (webknjaz) [inline MANIFEST.in:9]: Tests and everything related should be included in sdist because it's used as a source for downstream packaging normally. Besides, it doesn't really make much sense to maintain this exclusion list — just include everything from Git and forget about it. Drop this file altogether.

File: Makefile [hunk 3]
```diff
@@ -5,9 +5,6 @@ FLAGS=
 checkrst:
 	python setup.py check --restructuredtext
 
-pyroma:
```
Comment (webknjaz) [inline Makefile:8]: This could be in a separate PR.

File: Makefile [hunk 4]
```diff
@@ -5,9 +5,6 @@ FLAGS=
...
-	pyroma -d .
-
 
 flake:checkrst pyroma
```
Comment (webknjaz) [inline Makefile:9]: This `pyroma` dependency target seems forgotten here.

File: aiomysql/__init__.py [hunk 5]
```diff
@@ -23,6 +23,11 @@
 
 """
 
+try:
+    import pkg_resources
```
Comment (webknjaz) [inline aiomysql/__init__.py:27]: The use of `pkg_resources` is highly discouraged. Replace it with `importlib_metadata` (add an optional backport to the runtime deps if needed — expecting an import error is not necessary).

File: aiomysql/__init__.py [hunk 6]
```diff
@@ -33,7 +38,10 @@
 from .cursors import Cursor, SSCursor, DictCursor, SSDictCursor
 from .pool import create_pool, Pool
 
-__version__ = '0.0.22'
+try:
+    __version__ = pkg_resources.get_distribution('aiomysql').version
```
Comment (webknjaz) [inline aiomysql/__init__.py:42]: This approach is a bit outdated. I suggest using `setuptools-scm`'s `write_to` setting. We could have it generate a `_scm_version.py` file that is gitignored. And there could be a checked in `_version.py` that imports `version as __version__` from there and contains a fallback logic. This way, we could avoid having logic in the package init.

File: docs/conf.py [hunk 7]
```diff
@@ -78,13 +60,6 @@ def get_version(release):
 project = 'aiomysql'
 copyright = '2015,2016 Nikolay Novik'
 
-# The version info for the project you're documenting, acts as replacement for
-# |version| and |release|, also used in various other places throughout the
-# built documents.
-#
-release = get_release()
-version = get_version(release)
```
Comment (webknjaz) [inline docs/conf.py:86]: These are Sphinx's built-in settings, with these exact names. They should be kept. I suggest adding `setuptools-scm` as a docs build dependency. `get_version()` could be imported and generate both of these vars. The only extra thing to take care of is making sure to `git fetch --tags --all` in the local docs building wrapper. I like this integration in the docs because then I can show the precise commit sha in the build.

File: pyproject.toml [hunk 8]
```diff
@@ -0,0 +1,9 @@
+[build-system]
+requires = [
+    "setuptools>=42",
+    "setuptools_scm[toml]>=6.4",
+    "wheel",
```
Comment (webknjaz) [inline pyproject.toml:5]: `wheel` is auto-injected via the setuptools' PEP 517 build back-end. Also, there's a bit missing that allows tagged `pip install https://github//...git/archive.tar.gz` to work.
```suggestion
    # Essentials
    "setuptools >= 42",

    # Plugins
    "setuptools_scm[toml] >= 6.4",
    "setuptools_scm_git_archive >= 1.1",
```
For `setuptools_scm_git_archive` to work, don't forget to add `.git_archival.txt` containing:
```txt
ref-names: $Format:%D$
```
and `.gitattributes` with:
```gitattributes
# Force LF line endings for text files
* text=auto eol=lf

# Needed for setuptools-scm-git-archive
.git_archival.txt  export-subst
```

File: pyproject.toml [hunk 9]
```diff
@@ -0,0 +1,9 @@
...
+]
+build-backend = "setuptools.build_meta"
+
+[tool.setuptools_scm]
```
Comment (webknjaz) [inline pyproject.toml:9]: https://github.com/aio-libs/aiomysql/pull/734/files#r812397051 can be implemented via
```suggestion
[tool.setuptools_scm]
write_to = "aiomysql/_scm_version.py"
```
and `aiomysql/.gitignore` containing
```gitignore
/_scm_version.py
```
And to make up for missing typing, a pyi file could be created, like https://github.com/sphinx-contrib/sphinxcontrib-towncrier/blob/master/src/sphinxcontrib/towncrier/_scm_version.pyi.

File: setup.cfg [hunk 10]
```diff
@@ -2,6 +2,11 @@
 name = aiomysql
 url = https://github.com/aio-libs/aiomysql
 download_url = https://pypi.python.org/pypi/aiomysql
+project_urls =
```
Comment (webknjaz) [inline setup.cfg:5]: Mind putting this in a separate PR? + maybe add a changelog link too?

File: setup.cfg [hunk 11]
```diff
@@ -29,10 +34,12 @@ platforms =
   POSIX
 
 [options]
+use_scm_version = True
```
Comment (webknjaz) [inline setup.cfg:37]: This is legacy and doesn't really mean anything nowadays with the modern packaging toolchain.
```suggestion
```

File: setup.cfg [hunk 12]
```diff
@@ -41,3 +48,8 @@ sa =
   sqlalchemy>=1.0,<1.4
 rsa =
   PyMySQL[rsa]>=1.0
+
+[options.packages.find]
+exclude =
+  tests
+  tests.*
```
Comment (webknjaz) [inline setup.cfg:55]: Are you sure both entries are needed?

File: setup.py [hunk 13]
```diff
@@ -1,25 +1,11 @@
...
-                return match.group(1)
-        else:
-            raise RuntimeError('Cannot find version in aiomysql/__init__.py')
-
-
-setup(version=read_version(),
-      long_description='\n\n'.join((read('README.rst'), read('CHANGES.txt'))),
-      packages=find_packages(exclude=['tests', 'tests.*']))
+setup(
+    long_description='\n\n'.join((read('README.rst'), read('CHANGES.txt'))),
```
Comment (webknjaz) [inline setup.py:10]: Maybe just put `long_description = file: README.rst, CHANGES.txt` into `setup.cfg` and drop this file?

Also, this seems like a thing to do in a separate PR.

File: setup.py [hunk 14]
```diff
@@ -1,25 +1,11 @@
...
-    with open(init_py) as f:
-        for line in f:
-            match = regexp.match(line)
-            if match is not None:
-                return match.group(1)
-        else:
-            raise RuntimeError('Cannot find version in aiomysql/__init__.py')
-
-
-setup(version=read_version(),
```
Comment (webknjaz) [inline setup.py:23]: This could be implemented as `version = attr: aiomysql.__version__` in `setup.cfg`, in a dedicated PR that just drops `setup.py`. And setuptools-scm integration could be done in a follow-up. I just feel like there's a lot of things mixed together in this patch which makes it harder to reason about the effect of the changes.

File: setup.cfg [hunk 15]
```diff
@@ -2,6 +2,11 @@
 name = aiomysql
 url = https://github.com/aio-libs/aiomysql
 download_url = https://pypi.python.org/pypi/aiomysql
+project_urls =
+  CI: GitHub = https://github.com/aio-libs/aiomysql/actions
+  Docs: RTD = https://aiomysql.readthedocs.io/
+  GitHub: issues = https://github.com/aio-libs/aiomysql/issues
```
Comment (webknjaz) [inline setup.cfg:8]: Also, maybe link https://github.com/aio-libs/aiomysql/discussions too

File: requirements-dev.txt [hunk 16]
```diff
@@ -4,9 +4,8 @@ ipdb==0.13.9
 pytest==7.0.1
 pytest-cov==3.0.0
 pytest-sugar==0.9.4
-PyMySQL>=1.0.0,<=1.0.2
-sphinx>=1.8.1, <4.4.1
+PyMySQL==1.0.2
+Sphinx>=4.0
 sphinxcontrib-asyncio==0.3.0
-sqlalchemy>1.2.12,<=1.3.24
+sqlalchemy==1.3.24
```
Comment (webknjaz) [inline requirements-dev.txt:10]: Maybe move the unrelated mangling with pins to a separate PR too?

File: docs/conf.py (hunk 7 above)
Comment (webknjaz) [reply docs/conf.py:86]: How about doing the Sphinx + setuptools-scm integration into a completely separate effort?

File: docs/conf.py [hunk 17]
```diff
@@ -32,24 +32,6 @@
 
 import re, os.path
```
Comment (webknjaz) [inline docs/conf.py:33]: Is this still used?

File: .github/workflows/ci-cd.yml [hunk 18]
```diff
@@ -0,0 +1,754 @@
...
+        '${{ needs.pre-setup.outputs.git-tag }}'
+        --
+        ${{ github.event.inputs.release-commitish }}
+    - name: Build dists
+      run: >-
+        python
+        -m
+        build
+        --sdist
+        --wheel
```
Comment (webknjaz) [inline .github/workflows/ci-cd.yml:293]: It's best not to have these args explicit. If they are, both the sdist and the wheel are created from Git checkout. But if you omit them, then the wheel is created from the sdist which also makes sure to test how pip builds wheels.
```suggestion
```

File: .github/workflows/ci-cd.yml [hunk 19]
```diff
@@ -0,0 +1,754 @@
...
+        --user
+        --requirement requirements-dev.txt
+
+    - name: Install previously built wheel
+      run: >-
+        python -m
+        pip install
+        --no-cache-dir
+        --user
+        --no-index
```
Comment (webknjaz) [inline .github/workflows/ci-cd.yml:522]: I don't think these are needed. With an explicit file path, the wheel we need will get installed and the deps could be received from wherever they are.
```suggestion
        --user
```

File: .github/workflows/ci-cd.yml [hunk 20]
```diff
@@ -0,0 +1,754 @@
...
+        python -c
+        "import ssl; print('\nOPENSSL_VERSION: '
+        + ssl.OPENSSL_VERSION + '\nOPENSSL_VERSION_INFO: '
+        + repr(ssl.OPENSSL_VERSION_INFO)
+        + '\nOPENSSL_VERSION_NUMBER: '
+        + repr(ssl.OPENSSL_VERSION_NUMBER))"
+
+    - name: Check rst
+      run: |
+        python setup.py check --restructuredtext
```
Comment (webknjaz) [inline .github/workflows/ci-cd.yml:555]: With the suggestion to drop `setup.py`, I suggest replacing the linter that checks RST. After all, invoking `setup.py` is actively discouraged: https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html.

File: .github/workflows/ci-cd.yml [hunk 21]
```diff
@@ -0,0 +1,754 @@
...
+            mysql -h127.0.0.1 -uroot "-p$MYSQL_ROOT_PASSWORD" -e 'select version()' && break
+        done
+
+        mysql -h127.0.0.1 -uroot "-p$MYSQL_ROOT_PASSWORD" -e "SET GLOBAL local_infile=on"
+
+    - name: Run tests
+      run: |
+        # timeout ensures a more or less clean stop by sending a KeyboardInterrupt which will still provide useful logs
+        timeout --preserve-status --signal=INT --verbose 5m \
+          pytest --capture=no --verbosity 2 --cov-report term --cov-report xml --cov aiomysql --cov tests ./tests --mysql-unix-socket "unix-${{ join(matrix.db, '') }}=/tmp/run-${{ join(matrix.db, '-') }}/mysql.sock" --mysql-address "tcp-${{ join(matrix.db, '') }}=127.0.0.1:3306"
```
Comment (webknjaz) [inline .github/workflows/ci-cd.yml:600]: Consider putting the static options into `pytest.ini`. Or even wrap the invocation with `tox` to allow reusing the same args locally and in the CI. Maybe in a follow-up.

File: .github/workflows/ci-cd.yml [hunk 22]
```diff
@@ -0,0 +1,754 @@
...
+      env:
+        PYTHONUNBUFFERED: 1
+      timeout-minutes: 6
+
+    - name: Upload coverage
+      uses: codecov/codecov-action@v2.1.0
+      with:
+        file: ./coverage.xml
+        flags: "${{ matrix.os }}_${{ matrix.py }}_${{ join(matrix.db, '-') }}"
```
Comment (webknjaz) [inline .github/workflows/ci-cd.yml:609]: It's best to use comma-separated flags. Codecov allows ticking each one separately and combining to choose different aspects of coverage to show.

File: .github/workflows/ci-cd.yml [hunk 23]
```diff
@@ -0,0 +1,754 @@
...
+        ${{ needs.pre-setup.outputs.git-tag }}
+      uses: ncipollo/release-action@40bb172bd05f266cf9ba4ff965cb61e9ee5f6d01
+      with:
+        artifacts: |
+          dist/${{ needs.pre-setup.outputs.sdist-artifact-name }}
+          dist/${{ needs.pre-setup.outputs.wheel-artifact-name }}
+        artifactContentType: raw  # Because whl and tgz are of different types
+        # FIXME: Use Towncrier once it is integrated.
+        body: |
+          https://raw.githubusercontent.com/aio-libs/aiomysql/${{ needs.pre-setup.outputs.git-tag }}/CHANGES.txt
```
Comment (webknjaz) [inline .github/workflows/ci-cd.yml:751]: Maybe just include it?
```suggestion
        bodyFile: CHANGES.txt
```

File: Makefile (hunk 3 above)
Comment (Nothing4You) [reply Makefile:8]: moved to #737

File: Makefile (hunk 4 above)
Comment (Nothing4You) [reply Makefile:9]: addressed in #737

File: .github/workflows/ci-cd.yml (hunk 23 above)
Comment (Nothing4You) [reply .github/workflows/ci-cd.yml:751]: I was thinking about that, wasn't sure whether it would be a good idea to include the full changelog for all previous versions there.

File: setup.cfg (hunk 10 above)
Comment (Nothing4You) [reply setup.cfg:5]: created #738 to add the changelog to docs and then also reference it here

File: setup.cfg (hunk 11 above)
Comment (Nothing4You) [reply setup.cfg:37]: indeed, this is already implied by `[tool.setuptools_scm]` in `pyproject.toml`

File: setup.cfg (hunk 10 above)
Comment (Nothing4You) [reply setup.cfg:5]: moved this change to #739

File: setup.cfg (hunk 15 above)
Comment (Nothing4You) [reply setup.cfg:8]: done in #739

File: setup.cfg (hunk 12 above)
Comment (Nothing4You) [reply setup.cfg:55]: this was migrated from the current implementation, didn't review whether it's needed yet

File: .github/workflows/ci-cd.yml (hunk 20 above)
Comment (Nothing4You) [reply .github/workflows/ci-cd.yml:555]: that was migrated from the current implementation, should probably get replaced by `twine check`, but i didn't want to include that in this PR

File: .github/workflows/ci-cd.yml (hunk 21 above)
Comment (Nothing4You) [reply .github/workflows/ci-cd.yml:600]: this is also just migrated from the current implementation, tox is certainly something i still want to look into. out of scope of this PR though.

File: .github/workflows/ci-cd.yml (hunk 18 above)
Comment (Nothing4You) [reply .github/workflows/ci-cd.yml]: thanks, didn't realize it affected wheel creation

File: .github/workflows/ci-cd.yml (hunk 22 above)
Comment (Nothing4You) [reply .github/workflows/ci-cd.yml:607]: I'm not sure if I'm missing something here or it's just not implemented correctly in the `cheroot` repo?

https://github.com/cherrypy/cheroot/blob/81f9c687c264d0ec2c1b8bfa3b59caffe739db72/.github/workflows/ci-cd.yml#L689-L693

https://github.com/cherrypy/cheroot/runs/5147091997?check_suite_focus=true#step:19:4
logs
```yaml
with:
  flags: GHA, Linux, 3.4, cheroot-8.6.1.dev388+ge1fb8e54-py2.py3-none-any.whl
```
the upload URL then only contains `&flags=GHA`:
https://github.com/cherrypy/cheroot/runs/5147091997?check_suite_focus=true#step:19:38
and on codecov this only shows a single flag: https://app.codecov.io/gh/cherrypy/cheroot/compare/491/overview

while on the current configuration I can see a lot of flags on codecov:
https://app.codecov.io/gh/aio-libs/aiomysql/compare/706

Comment (Nothing4You) [reply .github/workflows/ci-cd.yml:607]: according to [codecov-action's arg docs](https://github.com/marketplace/actions/codecov#arguments)

> Flag the upload to group coverage metrics (unittests, uitests, etc.). Multiple flags are separated by a comma (ui,chrome)

considering https://docs.codecov.com/docs/flags#one-to-one-relationship-of-flags-to-uploads documents that flags should be 1:1, my understanding would be that the same flag should not be used by multiple uploads, unless it replaces the previous upload's data.

the example of `(ui,chrome)` does however make it sound reasonable to also use e.g. `(ui,firefox)`.

Comment (Nothing4You) [reply .github/workflows/ci-cd.yml:607]: The action definitely does pass the separated flags to the codecov executable though: https://github.com/cherrypy/cheroot/runs/5147091997?check_suite_focus=true#step:19:21

File: requirements-dev.txt (hunk 16 above)
Comment (Nothing4You) [reply requirements-dev.txt]: removed, partially split to #740

File: CHANGES.txt (hunk 1 above)
Comment (Nothing4You) [reply CHANGES.txt]: I want to reformat the changelog before the next release anyways to implement proper formatting and wording.

File: MANIFEST.in (hunk 2 above)
Comment (Nothing4You) [reply MANIFEST.in]: makes sense, removed

File: setup.cfg (hunk 12 above)
Comment (Nothing4You) [reply setup.cfg:55]: SQLAlchemy tests are in a separate `tests.sa` package, so both are indeed needed.

File: pyproject.toml (hunk 8 above)
Comment (Nothing4You) [reply pyproject.toml:5]: moved the PEP 517 migration (without setuptools-scm) to #746, explicit wheel dependency is removed there.

File: setup.py (hunk 13 above)
Comment (Nothing4You) [reply setup.py:10]: thanks, implemented in #746

File: setup.py (hunk 14 above)
Comment (Nothing4You) [reply setup.py:23]: thanks, implemented in #746

File: .github/workflows/ci-cd.yml (hunk 20 above)
Comment (Nothing4You) [reply .github/workflows/ci-cd.yml:551]: replaced by `twine check --strict` in #746

File: aiomysql/__init__.py (hunk 5 above)
Comment (Nothing4You) [reply aiomysql/__init__.py:27]: with `setuptools-scm`'s `write_to` this won't be needed

File: aiomysql/__init__.py (hunk 6 above)
Comment (Nothing4You) [reply aiomysql/__init__.py:42]: implemented in #748

File: docs/conf.py (hunk 7 above)
Comment (Nothing4You) [reply docs/conf.py:86]: moved to #748

File: pyproject.toml (hunk 8 above)
Comment (Nothing4You) [reply pyproject.toml:5]: implemented in #748

File: pyproject.toml (hunk 9 above)
Comment (Nothing4You) [reply pyproject.toml:9]: implemented in #748

File: docs/conf.py (hunk 17 above)
Comment (Nothing4You) [reply docs/conf.py:33]: removed in #748

File: .github/workflows/ci-cd.yml [hunk 24]
```diff
@@ -0,0 +1,754 @@
...
+        --requirement requirements-dev.txt
+
+    - name: Install previously built wheel
+      run: >-
+        python -m
+        pip install
+        --no-cache-dir
+        --user
+        --no-index
+        'dist/${{ needs.pre-setup.outputs.wheel-artifact-name }}'
```
[the remark this answers is not in the document]
Comment (Nothing4You) [reply .github/workflows/ci-cd.yml:533]: done

File: .github/workflows/ci-cd.yml (hunk 22 above)
Comment (webknjaz) [reply .github/workflows/ci-cd.yml:617]: _TL;DR this would work today._

That judgement wasn't right. The better workaround would've been removing spaces between flags, and having commas like `flag1,flag2,flag3`.
The case with `flag1, flag2, flag3` wasn't being processed correctly back when you made this PR. It was a legit Codecov bug. Recently, I've found more, like stripping trailing zeroes when flags could be interpreted as floats.
The said bugs have been fixed in the action and the uploader projects. Some by me (not my first time fixing their bugs!)
Here are the changes, if you're curious: https://github.com/codecov/uploader/issues/876 / https://github.com/codecov/uploader/pull/878 / https://github.com/codecov/codecov-action/pull/791.

This now works: https://github.com/cherrypy/cheroot/blob/70a5980/.github/workflows/ci-cd.yml#L713-L724. I've migrated to having prefixed flags after discovering that bug with float-ish inputs, before the fix got released. Though I think it's even better this way so I kept the approach.
Another nice bit is extracting the exact Python version number that `actions/setup-python` installed ­— it's better then relying on the matrix factors (which may even contain chars that aren't acceptable for flags). See https://github.com/cherrypy/cheroot/blob/70a5980/.github/workflows/ci-cd.yml#L555.
